### PR TITLE
Add progress bar for weight loading

### PR DESCRIPTION
**Pull request by rasbt**
Previously, when someone downloaded a new model via CLI

```bash
litgpt download microsoft/phi-2
```

or

```python
from litgpt import LLM
llm = LLM.load("microsoft/phi-2")
```

it was very verbose when loading the model weights (which is useful for debugging):

```
Processing checkpoints/microsoft/phi-2/model-00001-of-00002.bin
Loading 'model.embed_tokens.weight' into RAM
Loading 'model.layers.0.input_layernorm.bias' into RAM
Loading 'model.layers.0.input_layernorm.weight' into RAM
Loading 'model.layers.0.mlp.fc1.bias' into RAM
Loading 'model.layers.0.mlp.fc1.weight' into RAM
Loading 'model.layers.0.mlp.fc2.bias' into RAM
Loading 'model.layers.0.mlp.fc2.weight' into RAM
Loading 'model.layers.0.self_attn.dense.bias' into RAM
Loading 'model.layers.0.self_attn.dense.weight' into RAM
Loading 'model.layers.1.input_layernorm.bias' into RAM
Loading 'model.layers.1.input_layernorm.weight' into RAM
Loading 'model.layers.1.mlp.fc1.bias' into RAM
Loading 'model.layers.1.mlp.fc1.weight' into RAM
Loading 'model.layers.1.mlp.fc2.bias' into RAM
Loading 'model.layers.1.mlp.fc2.weight' into RAM
Loading 'model.layers.1.self_attn.dense.bias' into RAM
Loading 'model.layers.1.self_attn.dense.weight' into RAM
Loading 'model.layers.10.input_layernorm.bias' into RAM
Loading 'model.layers.10.input_layernorm.weight' into RAM
Loading 'model.layers.10.mlp.fc1.bias' into RAM
Loading 'model.layers.10.mlp.fc1.weight' into RAM
Loading 'model.layers.10.mlp.fc2.bias' into RAM
Loading 'model.layers.10.mlp.fc2.weight' into RAM
Loading 'model.layers.10.self_attn.dense.bias' into RAM
Loading 'model.layers.10.self_attn.dense.weight' into RAM
Loading 'model.layers.11.input_layernorm.bias' into RAM
Loading 'model.layers.11.input_layernorm.weight' into RAM
Loading 'model.layers.11.mlp.fc1.bias' into RAM
Loading 'model.layers.11.mlp.fc1.weight' into RAM
Loading 'model.layers.11.mlp.fc2.bias' into RAM
Loading 'model.layers.11.mlp.fc2.weight' into RAM
Loading 'model.layers.11.self_attn.dense.bias' into RAM
Loading 'model.layers.11.self_attn.dense.weight' into RAM
Loading 'model.layers.12.input_layernorm.bias' into RAM
Loading 'model.layers.12.input_layernorm.weight' into RAM
Loading 'model.layers.12.mlp.fc1.bias' into RAM
Loading 'model.layers.12.mlp.fc1.weight' into RAM
Loading 'model.layers.12.mlp.fc2.bias' into RAM
Loading 'model.layers.12.mlp.fc2.weight' into RAM
Loading 'model.layers.12.self_attn.dense.bias' into RAM
Loading 'model.layers.12.self_attn.dense.weight' into RAM
Loading 'model.layers.13.input_layernorm.bias' into RAM
Loading 'model.layers.13.input_layernorm.weight' into RAM
Loading 'model.layers.13.mlp.fc1.bias' into RAM
Loading 'model.layers.13.mlp.fc1.weight' into RAM
Loading 'model.layers.13.mlp.fc2.bias' into RAM
Loading 'model.layers.13.mlp.fc2.weight' into RAM
Loading 'model.layers.13.self_attn.dense.bias' into RAM
Loading 'model.layers.13.self_attn.dense.weight' into RAM
Loading 'model.layers.14.input_layernorm.bias' into RAM
Loading 'model.layers.14.input_layernorm.weight' into RAM
Loading 'model.layers.14.mlp.fc1.bias' into RAM
Loading 'model.layers.14.mlp.fc1.weight' into RAM
Loading 'model.layers.14.mlp.fc2.bias' into RAM
Loading 'model.layers.14.mlp.fc2.weight' into RAM
Loading 'model.layers.14.self_attn.dense.bias' into RAM
Loading 'model.layers.14.self_attn.dense.weight' into RAM
Loading 'model.layers.15.input_layernorm.bias' into RAM
Loading 'model.layers.15.input_layernorm.weight' into RAM
Loading 'model.layers.15.mlp.fc1.bias' into RAM
Loading 'model.layers.15.mlp.fc1.weight' into RAM
Loading 'model.layers.15.mlp.fc2.bias' into RAM
Loading 'model.layers.15.mlp.fc2.weight' into RAM
Loading 'model.layers.15.self_attn.dense.bias' into RAM
Loading 'model.layers.15.self_attn.dense.weight' into RAM
Loading 'model.layers.16.input_layernorm.bias' into RAM
Loading 'model.layers.16.input_layernorm.weight' into RAM
Loading 'model.layers.16.mlp.fc1.bias' into RAM
Loading 'model.layers.16.mlp.fc1.weight' into RAM
Loading 'model.layers.16.mlp.fc2.bias' into RAM
Loading 'model.layers.16.mlp.fc2.weight' into RAM
Loading 'model.layers.16.self_attn.dense.bias' into RAM
Loading 'model.layers.16.self_attn.dense.weight' into RAM
Loading 'model.layers.17.input_layernorm.bias' into RAM
Loading 'model.layers.17.input_layernorm.weight' into RAM
Loading 'model.layers.17.mlp.fc1.bias' into RAM
Loading 'model.layers.17.mlp.fc1.weight' into RAM
Loading 'model.layers.17.mlp.fc2.bias' into RAM
Loading 'model.layers.17.mlp.fc2.weight' into RAM
Loading 'model.layers.17.self_attn.dense.bias' into RAM
Loading 'model.layers.17.self_attn.dense.weight' into RAM
Loading 'model.layers.18.input_layernorm.bias' into RAM
Loading 'model.layers.18.input_layernorm.weight' into RAM
Loading 'model.layers.18.mlp.fc1.bias' into RAM
Loading 'model.layers.18.mlp.fc1.weight' into RAM
Loading 'model.layers.18.mlp.fc2.bias' into RAM
Loading 'model.layers.18.mlp.fc2.weight' into RAM
Loading 'model.layers.18.self_attn.dense.bias' into RAM
Loading 'model.layers.18.self_attn.dense.weight' into RAM
Loading 'model.layers.19.input_layernorm.bias' into RAM
Loading 'model.layers.19.input_layernorm.weight' into RAM
Loading 'model.layers.19.mlp.fc1.bias' into RAM
Loading 'model.layers.19.mlp.fc1.weight' into RAM
Loading 'model.layers.19.mlp.fc2.bias' into RAM
Loading 'model.layers.19.mlp.fc2.weight' into RAM
Loading 'model.layers.19.self_attn.dense.bias' into RAM
Loading 'model.layers.19.self_attn.dense.weight' into RAM
Loading 'model.layers.2.input_layernorm.bias' into RAM
Loading 'model.layers.2.input_layernorm.weight' into RAM
Loading 'model.layers.2.mlp.fc1.bias' into RAM
Loading 'model.layers.2.mlp.fc1.weight' into RAM
Loading 'model.layers.2.mlp.fc2.bias' into RAM
Loading 'model.layers.2.mlp.fc2.weight' into RAM
Loading 'model.layers.2.self_attn.dense.bias' into RAM
Loading 'model.layers.2.self_attn.dense.weight' into RAM
Loading 'model.layers.20.input_layernorm.bias' into RAM
Loading 'model.layers.20.input_layernorm.weight' into RAM
Loading 'model.layers.20.mlp.fc1.bias' into RAM
Loading 'model.layers.20.mlp.fc1.weight' into RAM
Loading 'model.layers.20.mlp.fc2.bias' into RAM
Loading 'model.layers.20.mlp.fc2.weight' into RAM
Loading 'model.layers.20.self_attn.dense.bias' into RAM
Loading 'model.layers.20.self_attn.dense.weight' into RAM
Loading 'model.layers.21.input_layernorm.bias' into RAM
Loading 'model.layers.21.input_layernorm.weight' into RAM
Loading 'model.layers.21.mlp.fc1.bias' into RAM
Loading 'model.layers.21.mlp.fc1.weight' into RAM
Loading 'model.layers.21.mlp.fc2.bias' into RAM
Loading 'model.layers.21.mlp.fc2.weight' into RAM
Loading 'model.layers.21.self_attn.dense.bias' into RAM
Loading 'model.layers.21.self_attn.dense.weight' into RAM
Loading 'model.layers.22.input_layernorm.bias' into RAM
Loading 'model.layers.22.input_layernorm.weight' into RAM
Loading 'model.layers.22.mlp.fc1.bias' into RAM
Loading 'model.layers.22.mlp.fc1.weight' into RAM
Loading 'model.layers.22.mlp.fc2.bias' into RAM
Loading 'model.layers.22.mlp.fc2.weight' into RAM
Loading 'model.layers.22.self_attn.dense.bias' into RAM
Loading 'model.layers.22.self_attn.dense.weight' into RAM
Loading 'model.layers.23.input_layernorm.bias' into RAM
Loading 'model.layers.23.input_layernorm.weight' into RAM
Loading 'model.layers.23.mlp.fc1.bias' into RAM
Loading 'model.layers.23.mlp.fc1.weight' into RAM
Loading 'model.layers.23.mlp.fc2.bias' into RAM
Loading 'model.layers.23.mlp.fc2.weight' into RAM
Loading 'model.layers.23.self_attn.dense.bias' into RAM
Loading 'model.layers.23.self_attn.dense.weight' into RAM
Loading 'model.layers.24.input_layernorm.bias' into RAM
Loading 'model.layers.24.input_layernorm.weight' into RAM
Loading 'model.layers.24.mlp.fc1.bias' into RAM
Loading 'model.layers.24.mlp.fc1.weight' into RAM
Loading 'model.layers.24.mlp.fc2.bias' into RAM
Loading 'model.layers.24.mlp.fc2.weight' into RAM
Loading 'model.layers.24.self_attn.dense.bias' into RAM
Loading 'model.layers.24.self_attn.dense.weight' into RAM
Loading 'model.layers.25.input_layernorm.bias' into RAM
Loading 'model.layers.25.input_layernorm.weight' into RAM
Loading 'model.layers.25.mlp.fc1.bias' into RAM
Loading 'model.layers.25.mlp.fc1.weight' into RAM
Loading 'model.layers.25.mlp.fc2.bias' into RAM
Loading 'model.layers.25.mlp.fc2.weight' into RAM
Loading 'model.layers.25.self_attn.dense.bias' into RAM
Loading 'model.layers.25.self_attn.dense.weight' into RAM
Loading 'model.layers.26.input_layernorm.bias' into RAM
Loading 'model.layers.26.input_layernorm.weight' into RAM
Loading 'model.layers.26.mlp.fc1.bias' into RAM
Loading 'model.layers.26.mlp.fc1.weight' into RAM
Loading 'model.layers.26.mlp.fc2.bias' into RAM
Loading 'model.layers.26.mlp.fc2.weight' into RAM
Loading 'model.layers.26.self_attn.dense.bias' into RAM
Loading 'model.layers.26.self_attn.dense.weight' into RAM
Loading 'model.layers.27.input_layernorm.bias' into RAM
Loading 'model.layers.27.input_layernorm.weight' into RAM
Loading 'model.layers.27.mlp.fc1.bias' into RAM
Loading 'model.layers.27.mlp.fc1.weight' into RAM
Loading 'model.layers.27.mlp.fc2.bias' into RAM
Loading 'model.layers.27.mlp.fc2.weight' into RAM
Loading 'model.layers.27.self_attn.dense.bias' into RAM
Loading 'model.layers.27.self_attn.dense.weight' into RAM
Loading 'model.layers.28.input_layernorm.bias' into RAM
Loading 'model.layers.28.input_layernorm.weight' into RAM
Loading 'model.layers.28.mlp.fc1.bias' into RAM
Loading 'model.layers.28.mlp.fc1.weight' into RAM
Loading 'model.layers.28.mlp.fc2.bias' into RAM
Loading 'model.layers.28.mlp.fc2.weight' into RAM
Loading 'model.layers.28.self_attn.dense.bias' into RAM
Loading 'model.layers.28.self_attn.dense.weight' into RAM
Loading 'model.layers.29.input_layernorm.bias' into RAM
Loading 'model.layers.29.input_layernorm.weight' into RAM
Loading 'model.layers.29.mlp.fc1.bias' into RAM
Loading 'model.layers.29.mlp.fc1.weight' into RAM
Loading 'model.layers.29.mlp.fc2.bias' into RAM
Loading 'model.layers.29.mlp.fc2.weight' into RAM
Loading 'model.layers.29.self_attn.dense.bias' into RAM
Loading 'model.layers.29.self_attn.dense.weight' into RAM
Loading 'model.layers.3.input_layernorm.bias' into RAM
Loading 'model.layers.3.input_layernorm.weight' into RAM
Loading 'model.layers.3.mlp.fc1.bias' into RAM
Loading 'model.layers.3.mlp.fc1.weight' into RAM
Loading 'model.layers.3.mlp.fc2.bias' into RAM
Loading 'model.layers.3.mlp.fc2.weight' into RAM
Loading 'model.layers.3.self_attn.dense.bias' into RAM
Loading 'model.layers.3.self_attn.dense.weight' into RAM
Loading 'model.layers.4.input_layernorm.bias' into RAM
Loading 'model.layers.4.input_layernorm.weight' into RAM
Loading 'model.layers.4.mlp.fc1.bias' into RAM
Loading 'model.layers.4.mlp.fc1.weight' into RAM
Loading 'model.layers.4.mlp.fc2.bias' into RAM
Loading 'model.layers.4.mlp.fc2.weight' into RAM
Loading 'model.layers.4.self_attn.dense.bias' into RAM
Loading 'model.layers.4.self_attn.dense.weight' into RAM
Loading 'model.layers.5.input_layernorm.bias' into RAM
Loading 'model.layers.5.input_layernorm.weight' into RAM
Loading 'model.layers.5.mlp.fc1.bias' into RAM
Loading 'model.layers.5.mlp.fc1.weight' into RAM
Loading 'model.layers.5.mlp.fc2.bias' into RAM
Loading 'model.layers.5.mlp.fc2.weight' into RAM
Loading 'model.layers.5.self_attn.dense.bias' into RAM
Loading 'model.layers.5.self_attn.dense.weight' into RAM
Loading 'model.layers.6.input_layernorm.bias' into RAM
Loading 'model.layers.6.input_layernorm.weight' into RAM
Loading 'model.layers.6.mlp.fc1.bias' into RAM
Loading 'model.layers.6.mlp.fc1.weight' into RAM
Loading 'model.layers.6.mlp.fc2.bias' into RAM
Loading 'model.layers.6.mlp.fc2.weight' into RAM
Loading 'model.layers.6.self_attn.dense.bias' into RAM
Loading 'model.layers.6.self_attn.dense.weight' into RAM
Loading 'model.layers.7.input_layernorm.bias' into RAM
Loading 'model.layers.7.input_layernorm.weight' into RAM
Loading 'model.layers.7.mlp.fc1.bias' into RAM
Loading 'model.layers.7.mlp.fc1.weight' into RAM
Loading 'model.layers.7.mlp.fc2.bias' into RAM
Loading 'model.layers.7.mlp.fc2.weight' into RAM
Loading 'model.layers.7.self_attn.dense.bias' into RAM
Loading 'model.layers.7.self_attn.dense.weight' into RAM
Loading 'model.layers.8.input_layernorm.bias' into RAM
Loading 'model.layers.8.input_layernorm.weight' into RAM
Loading 'model.layers.8.mlp.fc1.bias' into RAM
Loading 'model.layers.8.mlp.fc1.weight' into RAM
Loading 'model.layers.8.mlp.fc2.bias' into RAM
Loading 'model.layers.8.mlp.fc2.weight' into RAM
Loading 'model.layers.8.self_attn.dense.bias' into RAM
Loading 'model.layers.8.self_attn.dense.weight' into RAM
Loading 'model.layers.9.input_layernorm.bias' into RAM
Loading 'model.layers.9.input_layernorm.weight' into RAM
Loading 'model.layers.9.mlp.fc1.bias' into RAM
Loading 'model.layers.9.mlp.fc1.weight' into RAM
Loading 'model.layers.9.mlp.fc2.bias' into RAM
Loading 'model.layers.9.mlp.fc2.weight' into RAM
Loading 'model.layers.9.self_attn.dense.bias' into RAM
Loading 'model.layers.9.self_attn.dense.weight' into RAM
Loading 'layer 0 q bias' into RAM
Loading 'layer 0 k bias' into RAM
Loading 'layer 0 v bias' into RAM
Loading 'layer 0 q weight' into RAM
Loading 'layer 0 k weight' into RAM
Loading 'layer 0 v weight' into RAM
Loading 'layer 1 q bias' into RAM
Loading 'layer 1 k bias' into RAM
Loading 'layer 1 v bias' into RAM
Loading 'layer 1 q weight' into RAM
Loading 'layer 1 k weight' into RAM
Loading 'layer 1 v weight' into RAM
Loading 'layer 10 q bias' into RAM
Loading 'layer 10 k bias' into RAM
Loading 'layer 10 v bias' into RAM
Loading 'layer 10 q weight' into RAM
Loading 'layer 10 k weight' into RAM
Loading 'layer 10 v weight' into RAM
Loading 'layer 11 q bias' into RAM
Loading 'layer 11 k bias' into RAM
Loading 'layer 11 v bias' into RAM
Loading 'layer 11 q weight' into RAM
Loading 'layer 11 k weight' into RAM
Loading 'layer 11 v weight' into RAM
Loading 'layer 12 q bias' into RAM
Loading 'layer 12 k bias' into RAM
Loading 'layer 12 v bias' into RAM
Loading 'layer 12 q weight' into RAM
Loading 'layer 12 k weight' into RAM
Loading 'layer 12 v weight' into RAM
Loading 'layer 13 q bias' into RAM
Loading 'layer 13 k bias' into RAM
Loading 'layer 13 v bias' into RAM
Loading 'layer 13 q weight' into RAM
Loading 'layer 13 k weight' into RAM
Loading 'layer 13 v weight' into RAM
Loading 'layer 14 q bias' into RAM
Loading 'layer 14 k bias' into RAM
Loading 'layer 14 v bias' into RAM
Loading 'layer 14 q weight' into RAM
Loading 'layer 14 k weight' into RAM
Loading 'layer 14 v weight' into RAM
Loading 'layer 15 q bias' into RAM
Loading 'layer 15 k bias' into RAM
Loading 'layer 15 v bias' into RAM
Loading 'layer 15 q weight' into RAM
Loading 'layer 15 k weight' into RAM
Loading 'layer 15 v weight' into RAM
Loading 'layer 16 q bias' into RAM
Loading 'layer 16 k bias' into RAM
Loading 'layer 16 v bias' into RAM
Loading 'layer 16 q weight' into RAM
Loading 'layer 16 k weight' into RAM
Loading 'layer 16 v weight' into RAM
Loading 'layer 17 q bias' into RAM
Loading 'layer 17 k bias' into RAM
Loading 'layer 17 v bias' into RAM
Loading 'layer 17 q weight' into RAM
Loading 'layer 17 k weight' into RAM
Loading 'layer 17 v weight' into RAM
Loading 'layer 18 q bias' into RAM
Loading 'layer 18 k bias' into RAM
Loading 'layer 18 v bias' into RAM
Loading 'layer 18 q weight' into RAM
Loading 'layer 18 k weight' into RAM
Loading 'layer 18 v weight' into RAM
Loading 'layer 19 q bias' into RAM
Loading 'layer 19 k bias' into RAM
Loading 'layer 19 v bias' into RAM
Loading 'layer 19 q weight' into RAM
Loading 'layer 19 k weight' into RAM
Loading 'layer 19 v weight' into RAM
Loading 'layer 2 q bias' into RAM
Loading 'layer 2 k bias' into RAM
Loading 'layer 2 v bias' into RAM
Loading 'layer 2 q weight' into RAM
Loading 'layer 2 k weight' into RAM
Loading 'layer 2 v weight' into RAM
Loading 'layer 20 q bias' into RAM
Loading 'layer 20 k bias' into RAM
Loading 'layer 20 v bias' into RAM
Loading 'layer 20 q weight' into RAM
Loading 'layer 20 k weight' into RAM
Loading 'layer 20 v weight' into RAM
Loading 'layer 21 q bias' into RAM
Loading 'layer 21 k bias' into RAM
Loading 'layer 21 v bias' into RAM
Loading 'layer 21 q weight' into RAM
Loading 'layer 21 k weight' into RAM
Loading 'layer 21 v weight' into RAM
Loading 'layer 22 q bias' into RAM
Loading 'layer 22 k bias' into RAM
Loading 'layer 22 v bias' into RAM
Loading 'layer 22 q weight' into RAM
Loading 'layer 22 k weight' into RAM
Loading 'layer 22 v weight' into RAM
Loading 'layer 23 q bias' into RAM
Loading 'layer 23 k bias' into RAM
Loading 'layer 23 v bias' into RAM
Loading 'layer 23 q weight' into RAM
Loading 'layer 23 k weight' into RAM
Loading 'layer 23 v weight' into RAM
Loading 'layer 24 q bias' into RAM
Loading 'layer 24 k bias' into RAM
Loading 'layer 24 v bias' into RAM
Loading 'layer 24 q weight' into RAM
Loading 'layer 24 k weight' into RAM
Loading 'layer 24 v weight' into RAM
Loading 'layer 25 q bias' into RAM
Loading 'layer 25 k bias' into RAM
Loading 'layer 25 v bias' into RAM
Loading 'layer 25 q weight' into RAM
Loading 'layer 25 k weight' into RAM
Loading 'layer 25 v weight' into RAM
Loading 'layer 26 q bias' into RAM
Loading 'layer 26 k bias' into RAM
Loading 'layer 26 v bias' into RAM
Loading 'layer 26 q weight' into RAM
Loading 'layer 26 k weight' into RAM
Loading 'layer 26 v weight' into RAM
Loading 'layer 27 q bias' into RAM
Loading 'layer 27 k bias' into RAM
Loading 'layer 27 v bias' into RAM
Loading 'layer 27 q weight' into RAM
Loading 'layer 27 k weight' into RAM
Loading 'layer 27 v weight' into RAM
Loading 'layer 28 q bias' into RAM
Loading 'layer 28 k bias' into RAM
Loading 'layer 28 v bias' into RAM
Loading 'layer 28 q weight' into RAM
Loading 'layer 28 k weight' into RAM
Loading 'layer 28 v weight' into RAM
Loading 'layer 29 q bias' into RAM
Loading 'layer 29 k bias' into RAM
Loading 'layer 29 v bias' into RAM
Loading 'layer 29 q weight' into RAM
Loading 'layer 29 k weight' into RAM
Loading 'layer 29 v weight' into RAM
Loading 'layer 3 q bias' into RAM
Loading 'layer 3 k bias' into RAM
Loading 'layer 3 v bias' into RAM
Loading 'layer 3 q weight' into RAM
Loading 'layer 3 k weight' into RAM
Loading 'layer 3 v weight' into RAM
Loading 'layer 4 q bias' into RAM
Loading 'layer 4 k bias' into RAM
Loading 'layer 4 v bias' into RAM
Loading 'layer 4 q weight' into RAM
Loading 'layer 4 k weight' into RAM
Loading 'layer 4 v weight' into RAM
Loading 'layer 5 q bias' into RAM
Loading 'layer 5 k bias' into RAM
Loading 'layer 5 v bias' into RAM
Loading 'layer 5 q weight' into RAM
Loading 'layer 5 k weight' into RAM
Loading 'layer 5 v weight' into RAM
Loading 'layer 6 q bias' into RAM
Loading 'layer 6 k bias' into RAM
Loading 'layer 6 v bias' into RAM
Loading 'layer 6 q weight' into RAM
Loading 'layer 6 k weight' into RAM
Loading 'layer 6 v weight' into RAM
Loading 'layer 7 q bias' into RAM
Loading 'layer 7 k bias' into RAM
Loading 'layer 7 v bias' into RAM
Loading 'layer 7 q weight' into RAM
Loading 'layer 7 k weight' into RAM
Loading 'layer 7 v weight' into RAM
Loading 'layer 8 q bias' into RAM
Loading 'layer 8 k bias' into RAM
Loading 'layer 8 v bias' into RAM
Loading 'layer 8 q weight' into RAM
Loading 'layer 8 k weight' into RAM
Loading 'layer 8 v weight' into RAM
Loading 'layer 9 q bias' into RAM
Loading 'layer 9 k bias' into RAM
Loading 'layer 9 v bias' into RAM
Loading 'layer 9 q weight' into RAM
Loading 'layer 9 k weight' into RAM
Loading 'layer 9 v weight' into RAM
Processing checkpoints/microsoft/phi-2/model-00002-of-00002.bin
Loading 'lm_head.bias' into RAM
Loading 'lm_head.weight' into RAM
Loading 'model.final_layernorm.bias' into RAM
Loading 'model.final_layernorm.weight' into RAM
Loading 'model.layers.30.input_layernorm.bias' into RAM
Loading 'model.layers.30.input_layernorm.weight' into RAM
Loading 'model.layers.30.mlp.fc1.bias' into RAM
Loading 'model.layers.30.mlp.fc1.weight' into RAM
Loading 'model.layers.30.mlp.fc2.bias' into RAM
Loading 'model.layers.30.mlp.fc2.weight' into RAM
Loading 'model.layers.30.self_attn.dense.bias' into RAM
Loading 'model.layers.30.self_attn.dense.weight' into RAM
Loading 'model.layers.31.input_layernorm.bias' into RAM
Loading 'model.layers.31.input_layernorm.weight' into RAM
Loading 'model.layers.31.mlp.fc1.bias' into RAM
Loading 'model.layers.31.mlp.fc1.weight' into RAM
Loading 'model.layers.31.mlp.fc2.bias' into RAM
Loading 'model.layers.31.mlp.fc2.weight' into RAM
Loading 'model.layers.31.self_attn.dense.bias' into RAM
Loading 'model.layers.31.self_attn.dense.weight' into RAM
Loading 'layer 30 q bias' into RAM
Loading 'layer 30 k bias' into RAM
Loading 'layer 30 v bias' into RAM
Loading 'layer 30 q weight' into RAM
Loading 'layer 30 k weight' into RAM
Loading 'layer 30 v weight' into RAM
Loading 'layer 31 q bias' into RAM
Loading 'layer 31 k bias' into RAM
Loading 'layer 31 v bias' into RAM
Loading 'layer 31 q weight' into RAM
Loading 'layer 31 k weight' into RAM
Loading 'layer 31 v weight' into RAM
Saving converted checkpoint to checkpoints/microsoft/phi-2
```

This PR changes the default behavior to show progress bars instead:

```
Processing checkpoints/microsoft/phi-2/model-00001-of-00002.bin
Processing weights: 100%|██████████| 423/423 [00:13<00:00, 31.41it/s]
Processing QKV weights: 100%|██████████| 31/31 [00:00<00:00, 59.14it/s]
Processing checkpoints/microsoft/phi-2/model-00002-of-00002.bin
Processing weights: 100%|██████████| 30/30 [00:04<00:00,  7.04it/s]
Processing QKV weights: 100%|██████████| 32/32 [00:00<00:00, 870.16it/s]
Saving converted checkpoint to checkpoints/microsoft/phi-2
```

The old behavior can still be used via 

```bash
litgpt download microsoft/phi-2 --verbose true
```

CC @williamFalcon

### CODE REVIEW

**Comment by williamFalcon** [comment]
but why have 5 progress bars?

i'd prefer a single bar that measures ALL the progress but maybe prints a message saying what stage it's at (if needed).

it's like gettting 10 progress bars to start a laptop just because the laptop is doing 10 things under the hood.

**Comment by rasbt** [comment]
The problem is that we have n files and each file has m tensors. It's hard to now how many n*m tensors there are for a total progress bar until opening all the files and checking the number of weights. We can check this of course, but that would be adding minutes of extra time for large models.

We can maybe just have a progress bar over the m files instead.

**Comment by rasbt** [comment]
Ok figured out an efficient way to do a global progress bar.

Tests are failing because of a change in EleutherAI's GPTNeoX model in an external repo that we test against. Will investigate and fix in a separate PR and then revisit this.

**Comment by Andrei-Aksionov** [comment]
Maybe we could just retrieve the number of layers from the config file, monitor when the conversion script started working on another layer and use this as a signal to update the progress bar?

**Comment by rasbt** [comment]
There's now one global progress bar for all files and weights that are being loaded:

```
Loading weights: model-00003-of-00004.bin:  66%|███████████████████▎         | 00:22<00:10,  3.08it/s
```

File: litgpt/scripts/convert_hf_checkpoint.py [hunk 1]
```diff
@@ -4,22 +4,35 @@
...
 
+from tqdm import tqdm
 import torch
 from lightning.fabric.utilities.load import _NotYetLoadedTensor as NotYetLoadedTensor
 
 from litgpt import Config
 from litgpt.utils import extend_checkpoint_dir, incremental_save, lazy_load, save_config
 
 
+def estimate_total_entries(bin_files):
```
Comment (Andrei-Aksionov) [inline litgpt/scripts/convert_hf_checkpoint.py:20]: Is it used anywhere?

File: litgpt/scripts/convert_hf_checkpoint.py [hunk 2]
```diff
@@ -4,22 +4,35 @@
...
+    avg_param_size = 2 * 1024
+    return total_size // avg_param_size
+
+
 def copy_weights_gpt_neox(
     state_dict: Dict[str, torch.Tensor],
     hf_weights: Dict[str, Union[torch.Tensor, NotYetLoadedTensor]],
     saver: Optional[incremental_save] = None,
     dtype: Optional[torch.dtype] = None,
+    copy_fn: Optional[tqdm] = None,
```
Comment (Andrei-Aksionov) [inline litgpt/scripts/convert_hf_checkpoint.py:31]: I believe it's unused

File: litgpt/scripts/convert_hf_checkpoint.py (hunk 1 above)
Comment (rasbt) [reply litgpt/scripts/convert_hf_checkpoint.py:20]: Thanks, that's a leftover, I will clean this up.

File: litgpt/scripts/convert_hf_checkpoint.py [hunk 3]
```diff
@@ -394,11 +447,33 @@ def convert_hf_checkpoint(
...
+                    copy_fn(sd, hf_weights, saver=saver, dtype=dtype, pbar=pbar, progress_per_file=progress_per_file, debug_mode=debug_mode)
+                gc.collect()
+
+                if pbar.n < total_progress:
+                    pbar.update(total_progress - pbar.n)
+                pbar.close()
+        else:
+            # Handling files without progress bar in debug mode
+            for bin_file in sorted(bin_files):
+                current_file_size = os.path.getsize(bin_file)
```
Comment (Andrei-Aksionov) [inline litgpt/scripts/convert_hf_checkpoint.py:474]: This can be deleted

**Pull request merged**